### DISCUSSION TITLE
Property Annotation

### DISCUSSION THREAD

**Discussion by dezren39** [kdl-org/kdl#362] [Ideas]
Allow annotating properties in the same way that values and identifiers can be annotated.

**Comment by dezren39** [comment]
at least one implementation seems to have already implemented this

- https://pkg.go.dev/github.com/lunjon/gokdl#Prop
```
type Prop struct {
	Name  [string](https://pkg.go.dev/builtin#string)
	Value [any](https://pkg.go.dev/builtin#any)
	// ValueTypeAnnot is the type annotation for the value of the property.
	// Example: age=(u8)25
	// In this case it would be "u8".
	ValueTypeAnnot [TypeAnnotation](https://pkg.go.dev/github.com/lunjon/gokdl#TypeAnnotation)
	// TypeAnnot is the type annotation for the property itself.
	// Example: (author)name="Jonathan"
	TypeAnnot [TypeAnnotation](https://pkg.go.dev/github.com/lunjon/gokdl#TypeAnnotation)
}
```

**Comment by IceDragon200** [comment]
[Kuddle](https://github.com/IceDragon200/kuddle) supports property annotations (even in v1, which was a side effect of using the same Value struct for keys and values)

```elixir
defmodule Kuddle.Node do
  alias Kuddle.Value

  defstruct [
    name: nil,
    annotations: [],
    attributes: [],
    children: nil,
  ]

  @typedoc """
  An attribute is either a property `{key, value}`, or just a plain value.
  """
  @type attribute :: {key::Value.t(), value::Value.t()} | Value.t()

  @type t :: %__MODULE__{
    name: String.t(),
    annotations: [String.t()],
    attributes: [attribute()],
    children: [t()] | nil
  }
end

defmodule Kuddle.Value do
  defstruct [
    value: nil,
    annotations: [],
    type: :id,
    format: :plain,
  ]
end
```

```kdl
// This is legal in kuddle, but not to spec
(ntype)node (ktype)key=(vtype)"Something" 
```

I'd agree, allowing annotations on property keys should be fine, I can't think of a reason why it shouldn't be allowed.

**Comment by tabatkins** [comment]
I've got no particular problem with annotations on keys (they're trivial to parse, tho they do present some data-model complexities), but I do question their value. Key annotations would be similar to node annotations, rather than value annotations - they'll provide some additional context to the node/key, rather than telling you what kind of data is in the value. This is useful for nodes, because you can have multiple nodes with the same node name in a single child list, which are meant to provide different things (say, an `(author)person ...` vs a `(reviewer)person ...`). 

But key names have to be unique on a node already, so all necessary context is already baked in to their name. And the property's value can carry its own annotation when necessary.

So what is an actual use-case for a property key annotation?

**Reply by IceDragon200** [reply]
Personally, I've never needed it either, but it exists, the only thing I could think of is something like marking the key as a special type

```kdl
// EDIT: triggers (date)2023-12-18=send_to_billing (date)2023-12-19=check_billing
// The above was using an invalid identifier
triggers (date)"2023-12-18"=send_to_billing (date)"2023-12-19"=check_billing
```

But one could argue, that example is insane and would serve better as something like:

```kdl
triggers {
  trigger {
    (date)date 2023-12-18
    name send_to_billing
  }
  trigger {
    (date)date 2023-12-19
    name check_billing
  }
}
```

As you said it's trivial to parse, and as a proof of concept, kuddle parses it, but I can't think of a valid usecase where a better model wouldn't serve.

Still I think it's okay to implement, maybe someone will find a valid usecase for it one day

**Reply by tabatkins** [reply]
Your example is invalid since `2023-12-18` isn't a valid ident ^_^ And in general, yeah, this is def bad, you're trying to pretend that the key is a string value that can receive string annotations, but that doesn't extend at all (no other value types work there) and is kinda ridic.

**Reply by IceDragon200** [reply]
If I had taken even a second to run that through a proper parser, I would have noticed that, it's been a long day, I'll edit it for now

**Reply by dezren39** [reply]
my use case is that i want user provided files and am trying hard to get out of their way not tell them what to do. i want to be able to say
```
(final)Results (subject|official)reviewer=(name)Jon (review)type=in-person {
  Bob (pass)grade=86.2
  Sally (pass)grade=92.2
  Jim (fail)grade=12
}
```
these tags would be carried along with each piece of information and used for hinting/linking to other documents. i see the value mostly for dsl expressivity, i think there are times you want to tag what the key means, not the value, and a user isn't gonna care that the data model technically already covers that. a prefix method on keys like review-type review-name would be worse than property annotations. overloading value annotations for the entire key=val also sounds worse to me. rewriting it more in line with the regular way to write kdl sounds great, to me, and is what i might-- heavily suggest -- to users.. but i really don't want to tell them 'oh no the format wont let you' for an arbitrary (to them) reason.

**Reply by tabatkins** [reply]
Hm, can you elaborate on why you think `(review)type=in-person` is reasonable to write, but `review-type=in-person` isn't? Is the expectation that *sometimes* people will just write `type=in-person`, but they'll provide the tag if they wanna, I guess, comment it for themselves to make it more readable?

**Reply by dezren39** [reply]
basically, yeah. i would argue there's no reason to restrict it, and it can be useful for specific kinds of documents.

**Reply by tabatkins** [reply]
I generally disagree with that - we shouldn't be giving tools that don't actually do anything meaningful. While value annotations are well-justified, *node* annotations are already on much shakier ground, and that's *with* a few use-cases. If key annotations don't actually *do* anything meaningful, then we shouldn't add them, as they just become an attractive nuisance to users of the language. People will figure that, since they exist, there must be *some* use for them, and end up doing weird things just because they're available.

**Reply by IceDragon200** [reply]
@tabatkins After rewriting my reply around seven times, I decided to take a break and re-read the spec to understand your stance, and I hope by explaining that here, dezren39 can also understand why.

## Annotations 

Or rather [Type Annotations](https://github.com/kdl-org/kdl/blob/d53d99ff2e1519dab0bc9a636b863ed14288a05b/SPEC.md#type-annotation), while the spec makes lax assumptions about how these values should be interpreted, it is clear they are meant to represent some kind of data type, not so much as "tagging" as I had initially treated it, nothing is stopping anyone from using it for tagging however.

But its primary use is to denote the "type" of the item it is annotating (node or value), by that, I now understand why you said this:
> While value annotations are well-justified, node annotations are already on much shakier ground, and that's with a few use-cases. 

And to that, I present:
```kdl
node a b c
(list)node2 d
```

That's a node annotation used to inform the configuration that the attributes of the node should be treated as a list of items, for context [kuddle_config](https://github.com/IceDragon200/kuddle_config) by default treats _single_ arguments as immediate key value pairs:
```elixir
[
  node: ["a", "b", "c"],
  node2: ["d"]
]
```

Without the node annotation telling it to treat the second node as a list, it would be this:
```elixir
[
  node: ["a", "b", "c"],
  node2: "d"
]
```

You could argue: the value should have that annotation instead, except, in elixir a string > list is an actual operation that implies a charlist:
```kdl
node a b c
node2 (list)d
```

```elixir
[
  node: ["a", "b", "c"],
  node2: 'd' # or [100]
]
```

That justifies node annotations at least.

## Properties

The one thing I had completely overlooked is how [restrictive property keys](https://github.com/kdl-org/kdl/blob/d53d99ff2e1519dab0bc9a636b863ed14288a05b/SPEC.md#property) are, since they only allow strings (bare IDs or quoted effectively), which makes this invalid:

```kdl
node 1=2 3=4
```

However, there is a valid case for property annotations now:
```kdl
// This is a data_coding table for SMPP, the intention is to recode the left (key) to the right (value)
// In other words:
// GSM7 / Default, ASCII, and Latin 1 to UCS2 (UTF16)
recode (u32)"0"=8 (u32)"1"=8 (u32)"3"=8
```

This would have been better if properties allowed ANY value for its keys:
```kdl
recode 0=8 1=8 3=8
```

**Reply by tabatkins** [reply]
> they are meant to represent some kind of data type,

Right. Value annotations denote the value's type, with more precision than what you can get from just the syntax. Node annotations, well, *can* denote a type, but usually the node name will denote the type already, and so the node annotation instead serves as a key distinguishing it from other siblings of the same name. Thus examples like `(author)person ...` vs `(reviewer)person ...`. There are indeed example of using node annotations to denote types, of course; you gave one, and the official [JiK syntax](https://github.com/kdl-org/kdl/blob/main/JSON-IN-KDL.md) uses it as well to annotate ambiguous arrays or objects.

(Generally, values won't need this sort of "role" annotation, because you'd just make them properties in that case. The only remaining hole in the functionality is you can't tag multiple values with the same role, like you can nodes, unless you forgo using "type" annotations for them; but at that point the suggestion is to move to child nodes anyway.)

> However, there is a valid case for property annotations now:

That example requires your data model to be *very* specific, re-casting properties as simple pairs where the first value is required to be a string. It's arguable a minor abuse of KDL syntax, tho of course KDL is a tool that you can use however you find useful. I wouldn't want to change the language to *support* that kind of usage, however.
